Improve status bar

benmuth opened the pull request:
This PR tries to improve the status bar and status message handling in general. Resolves #21.

The previous status bar logic would create a new goroutine for every status message, leading to some undesired behavior if the function was called multiple times in quick succession.
I refactored the logic to use a buffered channel. Now any function that wants to display a status message just sends to the channel, and a separate goroutine receives from the channel and sets the editor to display the message for a given time. Subsequent messages wait for the prior message to finish displaying:

![pairpad-statusbar-1](https://user-images.githubusercontent.com/61804926/235547503-c30b9b9e-c722-4967-b574-d2dfaa8b3d5c.gif)

I also added a small persistent colored indicator to show whether you're connected to the server, just so it's always clear.

The status bar now shows the names of all active clients in different colors. It might be possible to later use each client's color to indicate their cursor once we implement multi-cursor support:

![pairpad-statusbar-2](https://user-images.githubusercontent.com/61804926/235547698-6495a1fc-800f-42c2-9a6c-5e0dfa777818.gif)

While working on this feature, the server crashed a few times due to an attempted concurrent write on a WebSocket connection. Following [this suggestion](https://github.com/gorilla/websocket/issues/119#issuecomment-198710015), I added a mutex around each read and write call in the server to avoid this.

benmuth committed:
Remove unused AddRune method.

The AddRune method was used for local insertions to update the
editor's slice of runes, bypassing the CRDT doc entirely. The
editor's slice of runes was then being immediately overwritten by
a CRDT insert, so AddRune was doing nothing.

This removes the AddRune method and slightly adapts the local
insertion logic to keep behavior the same.

benmuth committed:
Add mutex around local clock increment.

The GenerateInsert method increments the local clock. Without a  mutex,
a race condition was possible because both the local and message
handling threads were able to insert characters simultaneously.

benmuth committed:
Make status bar show active users.

Status bar now shows all active users connected to the server.

benmuth committed:
Add cell to status bar showing connection status.

benmuth committed:
Add mutex to avoid concurrent read/write panics.

The Gorilla WebSocket package does not support concurrent reads and
writes. The server would occasionally panic on a concurrent write, so
I followed the recommendation at (https://github.com/gorilla/websocket/
issues/119#issuecomment-198710015) to use a mutex.

Also changed the name of the `clientInfo` type to `client` to better
reflect its purpose.

benmuth committed:
Refactor status bar logic.

Status messages are now sent on a buffered channel, to allow for
multiple threads to update the status without conflict. There's a
separate status message handler that asynchronously listens on the
channel and temporarily shows any received messages.

Usernames in the status bar are now rendered in different colors.

benmuth committed:
Merge branch 'main' of https://github.com/burntcarrot/pairpad

benmuth committed:
Merge branch 'main' into statusbar

Bringing statusbar up to date with the upstream main branch.

benmuth committed:
Fix linter and DeepSource issues.

burntcarrot commented on client/editor/editor.go:
```diff
@@ -41,13 +40,39 @@ type Editor struct {
 
 // NewEditor returns a new instance of the editor.
 func NewEditor(conf EditorConfig) *Editor {
+	statusChan := make(chan string, 20)
```
Is there a specific reason for choosing `20` as the capacity? Perhaps it is because status updates are rare events and 20 is an adequate number to handle them?

benmuth replied:
It was an arbitrary choice, but that was my thinking, yeah. Though I don't really see any reason we shouldn't make it bigger. I think we basically never want the channel to block. Maybe a capacity of something like 100 would be more appropriate?

burntcarrot commented on client/editor/editor.go:
```diff
@@ -155,41 +181,58 @@ func (e *Editor) Draw() {
+	i := 0
+	for j, user := range e.Users {
+		for _, r := range user {
+			termbox.SetCell(i, e.Height-1, r, userColors[j], termbox.ColorDefault)
```
What happens when the number of users exceed the number of pre-defined user colors? I believe a modulus operator could help here (to loop back to zero)?

benmuth replied:
Ah of course, thank you!

burntcarrot commented on client/engine.go:
```diff
@@ -275,3 +279,19 @@ func getMsgChan(conn *websocket.Conn) chan commons.Message {
+
+		e.ShowMsg = true
+		e.Draw()
+		time.Sleep(3 * time.Second)
```
What is the effect introduced by `time.Sleep` here, as compared to the previous `time.AfterFunc`?

```go
// Toggle showMsg to false to hide the message.
	_ = time.AfterFunc(5*time.Second, func() {
		e.ShowMsg = false
	})
```

benmuth replied:
`time.AfterFunc` spawns a new goroutine for the function on every call, and the function was modifying the global `editor`. This worked if the `e.SetStatusBar` method was called once at a time, because the separate goroutine could handle the status bar state without blocking. If it was called multiple times (like to display multiple joining messages), some messages were never being shown because `e.ShowMsg` was set to false in one goroutine and never set back to true. Generally, the creation of multiple goroutines all modifying a global variable was hard to reason about.

I think the control flow here better matches the intended behavior of the status bar. Everything happens in one separate goroutine (called with `go handleStatusMsg()`), and within that goroutine everything happens in one thread: listen for a message, display the message for a given amount of time, then stop displaying the message, and repeat the cycle. All status bar modification happens in one thread rather than multiple.

burntcarrot reviewed (changes requested):
Looks good to me! ✅

Just a few changes:

- Could you please run the [Go race detector](https://go.dev/doc/articles/race_detector) and see if more changes are required?
- Could you update the places where we use `+` for string formatting to a more standard way like `fmt.Sprintf`?

benmuth committed:
Fix miscellaneous issues.

Changed string concatenation to use fmt.Sprintf() instead of '+'.
Fixed potential out-of-bounds error with the userColors slice.

benmuth commented:
Okay, for reference, here are the data races I detected for the client:

<details>

```
==================
WARNING: DATA RACE
Write at 0x00c0000ce140 by goroutine 13:
  main.handleStatusMsg()
      /Users/ben/Documents/Programming/pairpad/client/engine.go:290 +0x110

Previous read at 0x00c0000ce140 by main goroutine:
  github.com/burntcarrot/pairpad/client/editor.(*Editor).DrawStatusBar()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:192 +0x38
  github.com/burntcarrot/pairpad/client/editor.(*Editor).Draw()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:184 +0x3d4
  main.handleMsg()
      /Users/ben/Documents/Programming/pairpad/client/engine.go:252 +0x7c8
  main.mainLoop()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:55 +0xe0
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:31 +0x22c
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724

Goroutine 13 (running) created at:
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:29 +0x224
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724
==================
```


```
==================
WARNING: DATA RACE
Write at 0x000103191058 by goroutine 13:
  github.com/nsf/termbox-go.Clear()
      /Users/ben/go/pkg/mod/github.com/nsf/termbox-go@v1.1.1/api.go:453 +0x30
  github.com/burntcarrot/pairpad/client/editor.(*Editor).Draw()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:150 +0x30
  main.handleStatusMsg()
      /Users/ben/Documents/Programming/pairpad/client/engine.go:293 +0x15c

Previous write at 0x000103191058 by main goroutine:
  github.com/nsf/termbox-go.Clear()
      /Users/ben/go/pkg/mod/github.com/nsf/termbox-go@v1.1.1/api.go:453 +0x30
  github.com/burntcarrot/pairpad/client/editor.(*Editor).Draw()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:150 +0x30
  main.handleMsg()
      /Users/ben/Documents/Programming/pairpad/client/engine.go:252 +0x7c8
  main.mainLoop()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:55 +0xe0
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:31 +0x234
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724

Goroutine 13 (running) created at:
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:29 +0x22c
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724
==================
```


```
==================
WARNING: DATA RACE
Write at 0x00c000146068 by main goroutine:
  main.handleMsg()
      /Users/ben/Documents/Programming/pairpad/client/engine.go:220 +0x154
  main.mainLoop()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:62 +0xec
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:31 +0x2a0
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724

Previous read at 0x00c000146068 by goroutine 14:
  github.com/burntcarrot/pairpad/client/editor.(*Editor).DrawInfoBar()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:232 +0x38
  github.com/burntcarrot/pairpad/client/editor.(*Editor).DrawStatusBar()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:206 +0x70
  github.com/burntcarrot/pairpad/client/editor.(*Editor).Draw()
      /Users/ben/Documents/Programming/pairpad/client/editor/editor.go:192 +0x3d4
  main.mainLoop.func1()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:44 +0x74

Goroutine 14 (running) created at:
  main.mainLoop()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:42 +0x34
  main.initUI()
      /Users/ben/Documents/Programming/pairpad/client/ui.go:31 +0x2a0
  main.main()
      /Users/ben/Documents/Programming/pairpad/client/main.go:83 +0x724
==================
```
</details>

And for the server:

<details>

```
==================
WARNING: DATA RACE
Write at 0x00c000122cc0 by goroutine 7:
  runtime.mapaccessK()
      /usr/local/go/src/runtime/map.go:518 +0x1ec
  main.handleMsg()
      /Users/ben/Documents/Programming/pairpad/server/main.go:179 +0x16c

Previous read at 0x00c000122cc0 by goroutine 6:
  runtime.mapiterinit()
      /usr/local/go/src/runtime/map.go:815 +0x34c
  main.handleSync()
      /Users/ben/Documents/Programming/pairpad/server/main.go:221 +0x290

Goroutine 7 (running) created at:
  main.main()
      /Users/ben/Documents/Programming/pairpad/server/main.go:57 +0x130

Goroutine 6 (running) created at:
  main.main()
      /Users/ben/Documents/Programming/pairpad/server/main.go:54 +0x124
==================
==================
WARNING: DATA RACE
Write at 0x00c000194088 by goroutine 7:
  main.handleMsg()
      /Users/ben/Documents/Programming/pairpad/server/main.go:179 +0x178

Previous read at 0x00c000194088 by goroutine 6:
  main.handleSync()
      /Users/ben/Documents/Programming/pairpad/server/main.go:221 +0x2c0

Goroutine 7 (running) created at:
  main.main()
      /Users/ben/Documents/Programming/pairpad/server/main.go:57 +0x130

Goroutine 6 (running) created at:
  main.main()
      /Users/ben/Documents/Programming/pairpad/server/main.go:54 +0x124
==================
```
</details>

benmuth commented:
I'm close to done with a draft of the changes to address data races. Its scope goes beyond the status bar, and refactors a lot of the server logic and a good deal of the client logic as well. Do you think it would be a good idea for you to merge this PR so I can open a new one dedicated to the data race fixes?

benmuth committed:
Refactor to fix concurrency bugs and data races.

Previously, there were data races occurring due to multiple goroutines
concurrently reading from and writing to the `activeClients` map.

The main change is that there is now a monitor goroutine method called
`handle` that ensures that all modifications to the list of active
clients happen sequentially. Any other goroutine that wants to read
or write the list of active clients now has to send a request through
a channel, and the monitor goroutine selects a channel with a pending
request and fulfills the request.

I also refactored some of the main logic by splitting out the denser
sections of code into their own functions.

benmuth committed:
Fix race condition with accessing client names.

benmuth committed:
Fix client-side data races.

This is a first pass at addressing multiple data races on the client
side. The biggest change was serializing calls to Draw by sending on
a single channel to eliminate concurrency. Most other changes utilized
mutexes to protect concurrent read/write access of editor state.

I no longer detect data races at runtime during normal local use. More
rigorous detection could be done by increasing test coverage and running
the Go race detector with the test suite.

Not much consideration was given to performance. Profiling the client
and server code with the 'block' and 'mutex' memory profiles could point
to many areas for improvement.

burntcarrot committed:
ci: add race detector step
